File: apps/divsoup/lib/divsoup/analyzer/achievements/hyperlink_collector/silver.ex
defmodule Divsoup.Achievement.HyperlinkCollectorSilver do
  alias Divsoup.Achievement
  alias Divsoup.Util.Hyperlink

  @behaviour Divsoup.AchievementRule
  
  # Minimum number of unique external domains required
  @min_domains 10

  @impl true
  def evaluate(html_tree, _) do
    # Count unique external domains
    domain_count = Hyperlink.count_unique_external_domains(html_tree)
    
    if domain_count >= @min_domains do
      []
    else
      ["Page only links to #{domain_count} different external domains, needs at least #{@min_domains}"]
    end
  end

  @impl true
  def achievement() do
    %Achievement{
      hierarchy: :silver,
      title: "Hyperlink Custodian",
      group: "hyperlink_collector",
      description: "Page contains links to at least <strong>#{@min_domains}</strong> different external domains"
    }
  end
end

File: src/achievements/hyperlink_collector/silver.ts
import type { AchievementRule } from "../types";
import { evaluateRule } from "../evaluate";

export const rule: AchievementRule = {
  id: "hyperlink_collector.silver",
  title: "Hyperlink Custodian",
  group: "hyperlink_collector",
  description:
    "Page contains links to at least <strong>#{@min_domains}</strong> different external domains",
  hierarchy: "silver",
  evaluate: (context) => evaluateRule("hyperlink_collector.silver", context),
};
